src/Servers/Kestrel/Core/src/Internal/Http2/Http2Connection.cs

```diff
@@ -554,32 +563,57 @@ private Task ProcessHeadersFrameAsync<TContext>(IHttpApplication<TContext> appli
                 }
 
                 // Start a new stream
-                _currentHeadersStream = new Http2Stream<TContext>(application, new Http2StreamContext
-                {
-                    ConnectionId = ConnectionId,
-                    StreamId = _incomingFrame.StreamId,
-                    ServiceContext = _context.ServiceContext,
-                    ConnectionFeatures = _context.ConnectionFeatures,
-                    MemoryPool = _context.MemoryPool,
-                    LocalEndPoint = _context.LocalEndPoint,
-                    RemoteEndPoint = _context.RemoteEndPoint,
-                    StreamLifetimeHandler = this,
-                    ClientPeerSettings = _clientSettings,
-                    ServerPeerSettings = _serverSettings,
-                    FrameWriter = _frameWriter,
-                    ConnectionInputFlowControl = _inputFlowControl,
-                    ConnectionOutputFlowControl = _outputFlowControl,
-                    TimeoutControl = TimeoutControl,
-                });
-
-                _currentHeadersStream.Reset();
+                _currentHeadersStream = GetStream(application);
+
                 _headerFlags = _incomingFrame.HeadersFlags;
 
                 var headersPayload = payload.Slice(0, _incomingFrame.HeadersPayloadLength); // Minus padding
                 return DecodeHeadersAsync(_incomingFrame.HeadersEndHeaders, headersPayload);
             }
         }
 
+        private Http2Stream GetStream<TContext>(IHttpApplication<TContext> application)
+        {
+            if (_streamPool.TryPop(out var stream))
+            {
+                stream.Initialize(CreateHttp2StreamContext());
+                return stream;
+            }
+
+            return new Http2Stream<TContext>(
+                    application,
+                    CreateHttp2StreamContext());
+        }
+
+        private Http2StreamContext CreateHttp2StreamContext()
+        {
+            return new Http2StreamContext
+            {
+                ConnectionId = ConnectionId,
+                StreamId = _incomingFrame.StreamId,
```

Comment:
Could we use the same Http2StreamContext instance for all streams in a given connection by moving the StreamId out and making that a separate Http2Stream ctor?

Reply:
Is there some reason we couldn't do this?

Reply:
I did that! https://github.com/dotnet/aspnetcore/pull/18601/files#diff-259f4a7dbb73dd85ca5799a61ac2754fR576

Reply:
Only for pooled streams, which is an improvement. I was suggesting that the same Http2StreamContext be used for new streams as well, but nbd.